src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+                        knownState = Interlocked.Add(ref _state, modifyState) - modifyState;
+
+                        if (!waitSucceeded &&
+                            (knownState & LockStates.WriterSignaled) != 0 &&
+                            (knownState & LockStates.WaitingWritersMask) == LockStates.WaitingWriter)
+                        {
+                            if (writerEvent == null)
+                            {
+                                writerEvent = _writerEvent;
+                                Debug.Assert(writerEvent != null);
+                            }
+
+                            while (true)
+                            {
+                                knownState = _state;
+                                if ((knownState & LockStates.WriterSignaled) == 0 ||
+                                    (knownState & LockStates.WaitingWritersMask) != 0)
+                                {
+                                    break;
+                                }
+
+                                if (!writerEvent.WaitOne(10))
+                                {
+                                    continue;
+                                }
+
+                                modifyState = LockStates.Writer - LockStates.WriterSignaled;
+                                knownState = Interlocked.Add(ref _state, modifyState) - modifyState;
+                                Debug.Assert((knownState & LockStates.WriterSignaled) != 0);
+                                Debug.Assert((knownState & LockStates.Writer) == 0);
+
+                                // Honor the orginal status
+                                _writerID = threadID;
+                                Debug.Assert(_writerLevel == 0);
+                                _writerLevel = 1;
+                                ReleaseWriterLock();
+                                break;
+                            }
+                        }
+                    }
+
+                    if (!waitSucceeded)
+                    {
+                        throw GetTimeoutException();
+                    }
+                    break;
+                } while (YieldProcessor());
+            }
+
+            // Success
+            Debug.Assert((_state & LockStates.Writer) != 0);
+            Debug.Assert((_state & LockStates.ReadersMask) == 0);
+            Debug.Assert(_writerID == InvalidThreadID);
+
+            // Save threadid of the writer
+            _writerID = threadID;
+            _writerLevel = 1;
+            ++_writerSeqNum;
+#if RWLOCK_STATISTICS
+            ++_writerEntryCount;
+#endif
+            return;
+        }
+
+        public void AcquireWriterLock(TimeSpan timeout) => AcquireWriterLock(ToTimeoutMilliseconds(timeout));
+
+        public void ReleaseReaderLock()
+        {
+            // Check if the thread has writer lock
+            if (_writerID == GetCurrentThreadID())
+            {
+                ReleaseWriterLock();
```

Comment:
I guess it was decided at the time to have AcquireReaderLock acquire a write lock if the thread already owns a write lock. So code that expects to acquire a read lock, do work, and release the read lock, would instead acquire and release a recursive write lock.